drivers/accel/adxl355/adxl355.c

```diff
@@ -0,0 +1,934 @@
+
+	// Default value for POWER_CTL
+	dev->op_mode = (enum adxl355_op_mode)(GET_ADXL355_RESET_VAL(ADXL355_POWER_CTL));
+
+	// Default activity count value
+	dev->act_count = GET_ADXL355_RESET_VAL(ADXL355_ACT_COUNT);
+
+	*device = dev;
+
+	return ret;
+error_com:
+	if (dev->communication_type == ADXL355_SPI_COMM)
+		spi_remove(dev->spi_desc);
+	else
+		i2c_remove(dev->i2c_desc);
+	free(dev);
+	return ret;
+error_dev:
+	free(dev);
+	return ret;
+}
+
+/***************************************************************************//**
+ * @brief Free the resources allocated by adxl355_init().
+ *
+ * @param dev - The device structure.
+ *
+ * @return ret - Result of the remove procedure.
+*******************************************************************************/
+int adxl355_remove(struct adxl355_dev *dev)
+{
+	int ret;
+
+	if (dev->communication_type == ADXL355_SPI_COMM)
+		ret = spi_remove(dev->spi_desc);
+	else
+		ret = i2c_remove(dev->i2c_desc);
+
+	free(dev);
+
+	return ret;
+}
+
+/***************************************************************************//**
+ * @brief Reads from the device.
+ *
+ * @param dev          - The device structure.
+ * @param base_address - Address of the base register.
+ * @param size         - The number of bytes to be read and returned in read_data.
+ * @param read_data    - The read data buffer
+ *
+ * @return ret         - Result of the reading procedure.
+*******************************************************************************/
+int adxl355_read_device_data(struct adxl355_dev *dev, uint8_t base_address,
+			     uint16_t size, uint8_t *read_data)
+{
+	int ret;
+
+	if (dev->communication_type == ADXL355_SPI_COMM) {
+		uint8_t *comm_buffer;
+		comm_buffer = (uint8_t *)calloc(1,1 + size);
+		if (!comm_buffer)
+			return -ENOMEM;
+
+		comm_buffer[0] = ADXL355_SPI_READ | (base_address << 1);
+		ret = spi_write_and_read(dev->spi_desc, comm_buffer, 1 + size);
+		for (uint16_t idx = 0; idx < size; idx++)
+			read_data[idx] = comm_buffer[idx+1];
+
+		free(comm_buffer);
+	} else {
+		ret = i2c_write(dev->i2c_desc, &base_address, 1, 0);
+		if (ret)
+			return ret;
+		ret = i2c_read(dev->i2c_desc, read_data, size, 1);
+	}
+
+	return ret;
+}
+
+/***************************************************************************//**
+ * @brief Writes to the device
+ *
+ * @param dev          - The device structure.
+ * @param base_address - Address of the base register.
+ * @param size         - The number of bytes to be written. It is the size
+ *                       of the write_data buffer.
+ * @param write_data   - The data which is going to be written.
+ *
+ * @return ret         - Result of the writing procedure.
+*******************************************************************************/
+int adxl355_write_device_data(struct adxl355_dev *dev, uint8_t base_address,
+			      uint16_t size, uint8_t *write_data)
+{
+	int ret;
+	uint8_t *comm_buffer;
+	comm_buffer = (uint8_t *)calloc(1,1 + size);
```

Comment:
i think we should work directly on user buffer and not create the temporary buffer using malloc which can only cause troubles with a function like this one that may be called in critical code areas like intensive loops or interrupts

user is responsible for providing a big enough buffer

same comment for adxl355_read_device_data